sdk/communication/azure-communication-chat/azure/communication/chat/_shared/user_credential_async.py

```diff
@@ -3,93 +3,134 @@
 # Licensed under the MIT License. See License.txt in the project root for
 # license information.
 # --------------------------------------------------------------------------
+
 from asyncio import Condition, Lock
 from datetime import timedelta
-from typing import (  # pylint: disable=unused-import
-    cast,
-    Tuple,
-    Any
-)
-
+from typing import Any
+import sys
+import six
 from .utils import get_current_utc_as_int
-from .user_token_refresh_options import CommunicationTokenRefreshOptions
+from .utils import create_access_token
+from .utils_async import AsyncTimer
 
 
 class CommunicationTokenCredential(object):
     """Credential type used for authenticating to an Azure Communication service.
-    :param str token: The token used to authenticate to an Azure Communication service
-    :keyword token_refresher: The async token refresher to provide capacity to fetch fresh token
-    :raises: TypeError
+    :param str token: The token used to authenticate to an Azure Communication service.
+    :keyword token_refresher: The async token refresher to provide capacity to fetch a fresh token.
+     The returned token must be valid (expiration date must be in the future).
+    :paramtype token_refresher: Callable[[], Awaitable[AccessToken]]
+    :keyword bool proactive_refresh: Whether to refresh the token proactively or not.
+    :raises: TypeError if paramater 'token' is not a string
+    :raises: ValueError if the 'proactive_refresh' is enabled without providing the 'token_refresher' function.
     """
 
     _ON_DEMAND_REFRESHING_INTERVAL_MINUTES = 2
+    _DEFAULT_AUTOREFRESH_INTERVAL_MINUTES = 10
 
     def __init__(self, token: str, **kwargs: Any):
-        token_refresher = kwargs.pop('token_refresher', None)
-        communication_token_refresh_options = CommunicationTokenRefreshOptions(token=token,
-                                                                               token_refresher=token_refresher)
-        self._token = communication_token_refresh_options.get_token()
-        self._token_refresher = communication_token_refresh_options.get_token_refresher()
-        self._lock = Condition(Lock())
+        if not isinstance(token, six.string_types):
+            raise TypeError("Token must be a string.")
+        self._token = create_access_token(token)
+        self._token_refresher = kwargs.pop('token_refresher', None)
+        self._proactive_refresh = kwargs.pop('proactive_refresh', False)
+        if(self._proactive_refresh and self._token_refresher is None):
+            raise ValueError("'token_refresher' must not be None.")
+        self._timer = None
+        self._async_mutex = Lock()
+        if sys.version_info[:3] == (3, 10, 0):
+            # Workaround for Python 3.10 bug(https://bugs.python.org/issue45416):
+            getattr(self._async_mutex, '_get_loop', lambda: None)()
```

Comment:
Interesting bug!
I'm a _tiny_ bit concerned that the issue report mentions the fix is compatible with 3.7+, given that we're still tied to 3.6, we should make sure to test it :)

Reply:
I ran user credential samples and tests on 3.6.12 locally and works fine